packages/messaging/src/controllers/window-controller.ts

```diff
@@ -53,30 +67,137 @@ export class WindowController implements FirebaseMessaging, FirebaseService {
+  }): Promise<string> {
     if (Notification.permission === 'default') {
-      // The user hasn't allowed or denied notifications yet. Ask them.
       await Notification.requestPermission();
     }
 
     if (Notification.permission !== 'granted') {
       throw ERROR_FACTORY.create(ErrorCode.PERMISSION_BLOCKED);
     }
 
-    return getToken(this.firebaseDependencies, swRegistration, this.vapidKey);
+    await this.updateVapidKey(options?.vapidKey);
+    await this.updateSwReg(options?.serviceWorkerRegistration);
+
+    if (!this.swRegistration) {
+      console.debug(
+        TAG +
+          'no sw has been provided explicitly. Attempting to find firebase-messaging-sw.js in default directory.'
+      );
+      await this.registerDefaultSw();
+    }
+
+    return getToken(
+      this.firebaseDependencies,
+      this.swRegistration!,
+      this.vapidKey!
+    );
+  }
+
+  async updateVapidKey(vapidKey: string | undefined): Promise<void> {
+    if (!!this.vapidKey && !!vapidKey && this.vapidKey !== vapidKey) {
+      console.debug(
+        TAG +
+          'newly provided VapidKey is different from previously stored VapidKey.  New VapidKey is overriding.'
+      );
+      this.vapidKey = vapidKey;
+    }
+
+    if (!this.vapidKey && !!vapidKey) {
+      this.vapidKey = vapidKey;
+    }
+
+    if (!this.vapidKey && !vapidKey) {
+      console.debug(
+        TAG +
+          'no VapidKey is provided. Using the default VapidKey. Note that Push will NOT work in Chrome without a non-default VapidKey.'
+      );
+      this.vapidKey = DEFAULT_VAPID_KEY;
+    }
+  }
+
+  async updateSwReg(
+    swRegistration: ServiceWorkerRegistration | undefined
+  ): Promise<void> {
+    if (!swRegistration) {
+      return;
+    }
+
+    if (!(swRegistration instanceof ServiceWorkerRegistration)) {
+      throw ERROR_FACTORY.create(ErrorCode.INVALID_SW_REGISTRATION);
+    }
+
+    this.swRegistration = swRegistration;
+  }
+
+  private async registerDefaultSw(): Promise<void> {
+    try {
+      this.swRegistration = await navigator.serviceWorker.register(
+        DEFAULT_SW_PATH,
+        {
+          scope: DEFAULT_SW_SCOPE
+        }
+      );
+
+      // The timing when browser updates sw when sw has an update is unreliable by my experiment.
+      // It leads to version conflict when the SDK upgrades to a newer version in the main page, but
+      // sw is stuck with the old version. For example, https://github.com/firebase/firebase-js-sdk/issues/2590
+      // The following line reliably updates sw if there was an update.
+      this.swRegistration.update().catch(() => {
+        /* it is non blocking and we don't care if it failed */
+      });
+    } catch (e) {
+      throw ERROR_FACTORY.create(ErrorCode.FAILED_DEFAULT_REGISTRATION, {
+        browserErrorMessage: e.message
+      });
+    }
   }
 
   async deleteToken(): Promise<boolean> {
-    const swRegistration = await this.getServiceWorkerRegistration();
+    if (!this.swRegistration) {
```

Comment:
Should it be no-op if there is no swRegistration?

Reply:
It seems we can make `swRegistration` an optional parameter for `deleteToken()` (the imported function), so it will only try to unsubscribe if the registration is available. This way we don't have to register the default SW here.